debezium-core/src/main/java/io/debezium/metadata/ConnectorDescriptor.java

```diff
@@ -33,4 +41,38 @@ public String getClassName() {
     public String getVersion() {
         return version;
     }
+
+    public static String getIdForConnectorClass(String className) {
+        switch (className) {
+            case "io.debezium.connector.mongodb.MongoDbConnector":
+                return "mongodb";
+            case "io.debezium.connector.mysql.MySqlConnector":
+                return "mysql";
+            case "io.debezium.connector.oracle.OracleConnector":
+                return "oracle";
+            case "io.debezium.connector.postgresql.PostgresConnector":
+                return "postgres";
+            case "io.debezium.connector.sqlserver.SqlServerConnector":
+                return "sqlserver";
+            default:
+                throw new RuntimeException("Unsupported connector type with className: \"" + className + "\"");
+        }
+    }
```

Comment:
@jcechace Can you explain why this is a codesmell for you? There must be a place where we map from the Connector classname to a unique identifier for the UI. This is shared code used for a) the schema genarator, b) the Connect REST extensions, etc. We won't have the classes / classnames until we make up a dependency on all possible connector modules neither is there any other better place to describe that mapping from a classname to a connector type id. It belongs to what a ConnectorDescriptor defines / produces, so it should belong here until you suggest a better place for it. Without providing an alternative complaining is pointless, isn't it?

Reply:
There are methods to do it cleanly no? E.g. you extract this from conenctor config or via some declarative metadata such as annotation. You can even employ a design pattern such as strategy or whatnot.  

I understand why this is here -- due to it being a shared code. I'm saying that easiest to do (and incorrect) approach was taken here. I'm fine with this being a temporary solution but it is a code smell.

Reply:
Hi @rk3rn3r, when I typically see a large if-block or a switch statement, the first thing I usually ask is whether or not this could be refactored to take advantage of OOO design principles.

Something to consider might be whether you could take advantage of the ServiceLoader pattern and have a contract that each connector implements that exposes the connector class, its display name, and other metadata.  If you are deploying the REST extension alongside the connector as it seems you're doing with the connector-specific classes, this would work as there would only ever be a single REST implementation per connector plugin, no?

This would also work to avoid the need to deal with passing class names for connectors that are not supported and therefore having to throw exceptions like the reference Jakub shared.

Not saying this needs to change, but it's really something to consider to make code fundamentally more reliable and less more resilient to potential corner case failures.

Reply:
> Are you sure this will work in the way Kafka Connect REST Extensions are loaded? KC uses an isolated ClassLoader that should not be capable of finding artifacts in the connect plugin dirs of KC. I think this restriction then applies to ServiceLoader as well, doesn't it?

Well, it depends on how we intend to bundle the REST extension.  For example, if the extension is bundled with each connector plug-in, then the ClassLoader would only find the single implementation of the service for that specific connector, which I believe should satisfy API calls into the connector-specific REST endpoints. 

> Furthermore, this does not solve the issue that I described above, that the UI needs a stable identifier for connectors. If a connector can decide to change its "connector type id" then this is not stable. I think this code belongs into debezium-core or a similar central package in a way that defines the name globally instead of decentralized in the connector modules. wdyt?

> Furthermore, this does not solve the issue that I described above, that the UI needs a stable identifier for connectors. If a connector can decide to change its "connector type id" then this is not stable. I think this code belongs into debezium-core or a similar central package in a way that defines the name globally instead of decentralized in the connector modules. wdyt?

I disagree, I actually think something like this probably makes sense in each connector's `Module` implementation.  If you take a look at the connector configuration classes, you'll see that we have two methods that get exposed, `getContextName()` and `getConnectorName()`, and both of these are delegating to `Module`.  It's implied that we should not change these values as it could have implications since these are used in metrics bean names, the source info block, and so on.  

I could imagine `Module` being adopted as follows and perhaps one of the REST classes you load via ServiceLoader could then simply act like the configuration class and delegate the call to the local connector's `Module`.
```
public static <T> Class<T> getConnectorClass() {
  return OracleConnector.class;
}

public static String getConnectorClassName() {
  return getConnectorClass().getName();
}
```

> this applies in the same way to the SchemaGenerator which requires a stable connector type id for connectors
> The definition of these ids should be closer to the UI and SchemaGenerator than to the connector modules.

I don't see how that precludes anything that was suggested.  In fact, you already use ServiceLoader to fetch the specific `ConnectorMetadata` for the connector in question, so that same instance could have methods that delegate to the `Module` within each connector module to return that specific detail without the need for the switch statement.

In fact, if `ConnectorMetadata` delegated to `Module`, you would replace the hard-coded strings you have inside each of the metadata classes with a singular string reference that is used for a variety of reasons.  This means that if there is ever a reason to change the identifier in the future, this is managed in a singular place rather than across a multitude of files littered throughout the repository.

What do you think of this?  @jpechane @jcechace any takes?

Reply:
ConnectorMetadata should be removed eventually. It can only be used in the SchemaGenerator use-case but not from the REST extension side. We should migrate to this mapping here instead to produce stable identifiers for the UI frontend / API side of things.

I filed https://issues.redhat.com/browse/DBZ-6950 for this to follow-up on the topic.